File: apps/web/src/types/strategies.ts
export type StrategyParameterType =
  "integer" | "decimal" | "boolean" | "string" | "enum";

export interface StrategyParameterDefinition {
  name: string;
  type: StrategyParameterType;
  required: boolean;
  default: string | number | boolean | null;
  description: string;
  min_value: string | number | null;
  max_value: string | number | null;
  choices: string[];
}

export interface StrategyCatalogItem {
  strategy_key: string;
  display_name: string;
  description: string;
  version: string;
  supported_timeframes: string[];
  parameter_schema_version: number;
  parameters: StrategyParameterDefinition[];
}

export interface StrategyRun {
  run_id: string;
  strategy_key: string;
  strategy_version: string;
  status: "CREATED" | "RUNNING" | "COMPLETED" | "FAILED";
  timeframe: string;
  instrument_ids: string[];
  instruments?: {
    id: string;
    symbol: string;
    exchange: string;
    name: string;
  }[];
  parameters?: Record<string, string | number | boolean>;
  start_at: string;
  end_at: string;
  bars_processed: number;
  signals_generated: number;
  warnings: string[];
  replayed?: boolean;
  started_at?: string | null;
  completed_at?: string | null;
  failed_at?: string | null;
  created_at?: string;
  error?: { code: string; message: string } | null;
  capabilities?: Record<string, boolean>;
}

export interface StrategyRunPage {
  items: StrategyRun[];
  page: number;
  page_size: number;
  total: number;
}

export interface StrategySignal {
  signal_id: string;
  strategy_run_id: string;
  sequence_number: number;
  strategy_key: string;
  strategy_version: string;
  instrument_id: string;
  instrument: {
    symbol: string | null;
    exchange: string | null;
    name: string | null;
  };
  signal_type: string;
  side: string;
  generated_at: string;
  bar_timestamp: string;
  quantity: string | null;
  target_weight: string | null;
  reference_price: string | null;
  confidence: string | null;
  reason: string | null;
  schema_version: number;
}

export interface StrategySignalPage {
  items: StrategySignal[];
  page: number;
  page_size: number;
  total: number;
}

export type ParameterGridValue = string | number | boolean;

export type StrategyExperimentStatus =
  "CREATED" | "RUNNING" | "COMPLETED" | "PARTIAL_FAILED" | "FAILED";

export interface StrategyExperimentSummary {
  experiment_id: string;
  idempotency_key: string;
  strategy_key: string;
  strategy_version: string;
  environment: string;
  timeframe: string;
  instrument_ids: string[];
  start_at: string;
  end_at: string;
  parameter_grid: Record<string, ParameterGridValue[]>;
  combination_count: number;
  runs_completed: number;
  runs_failed: number;
  total_signals: number;
  status: StrategyExperimentStatus;
  started_at: string | null;
  completed_at: string | null;
  failed_at: string | null;
  error: { code: string; message: string } | null;
  correlation_id: string;
  created_at: string;
  updated_at: string;
  replayed?: boolean;
  capabilities?: Record<string, boolean>;
}

export type StrategyExperiment = StrategyExperimentSummary;

export interface StrategyExperimentPage {
  items: StrategyExperimentSummary[];
  page: number;
  page_size: number;
  total: number;
}

export interface StrategyExperimentRun {
  combination_index: number;
  normalized_parameters: Record<string, ParameterGridValue>;
  strategy_run_id: string | null;
  run_status: StrategyRun["status"];
  bars_processed: number;
  signals_generated: number;
  warning: string | null;
}

export interface StrategyExperimentComparisonRow {
  combination_index: number;
  normalized_parameters: Record<string, ParameterGridValue>;
  strategy_run_id: string | null;
  run_status: StrategyRun["status"];
  bars_processed: number;
  total_signals: number;
  buy_signals: number;
  sell_signals: number;
  first_signal_at: string | null;
  last_signal_at: string | null;
  signaled_instrument_count: number;
  warning: string | null;
}

export interface StrategySignalOverlap {
  left_combination_index: number;
  right_combination_index: number;
  intersection_count: number;
  union_count: number;
  similarity: string;
}

export interface CreateStrategyExperimentRequest {
  strategy_key: string;
  instrument_ids: string[];
  timeframe: string;
  start_at: string;
  end_at: string;
  parameter_grid: Record<string, ParameterGridValue[]>;
  idempotency_key: string;
}

File: apps/api/src/alphadesk_api/schemas/strategies.py
"""S01-C public strategy research API contracts."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from alphadesk_domain.market_reference import PriceAdjustmentMode


class StrategyParameterResponse(BaseModel):
    name: str
    type: str
    required: bool
    default: str | int | bool | None
    description: str
    min_value: str | int | None
    max_value: str | int | None
    choices: list[str]


class StrategyCatalogResponse(BaseModel):
    strategy_key: str
    display_name: str
    description: str
    version: str
    supported_timeframes: list[str]
    parameter_schema_version: int
    parameters: list[StrategyParameterResponse]


class StrategyRunCreateBody(BaseModel):
    strategy_key: str = Field(min_length=2, max_length=64)
    parameters: dict[str, str | int | bool] = Field(default_factory=dict)
    instrument_ids: list[UUID] = Field(min_length=1, max_length=100)
    timeframe: str
    start_at: datetime
    end_at: datetime
    idempotency_key: str = Field(min_length=1, max_length=128)
    price_adjustment_mode: PriceAdjustmentMode = PriceAdjustmentMode.RAW


class StrategyRunResponse(BaseModel):
    run_id: UUID
    strategy_key: str
    strategy_version: str
    status: str
    timeframe: str
    instrument_ids: list[UUID]
    start_at: datetime
    end_at: datetime
    bars_processed: int
    signals_generated: int
    warnings: list[str]
    replayed: bool = False
    price_adjustment_mode: PriceAdjustmentMode = PriceAdjustmentMode.RAW


class StrategyRunDetailResponse(StrategyRunResponse):
    parameters: dict[str, str | int | bool]
    instruments: list[dict[str, Any]]
    started_at: datetime | None
    completed_at: datetime | None
    failed_at: datetime | None
    created_at: datetime
    error: dict[str, str] | None
    capabilities: dict[str, bool]


class StrategyRunPageResponse(BaseModel):
    items: list[StrategyRunDetailResponse]
    page: int
    page_size: int
    total: int


class StrategySignalResponse(BaseModel):
    signal_id: UUID
    strategy_run_id: UUID
    sequence_number: int
    strategy_key: str
    strategy_version: str
    instrument_id: UUID
    instrument: dict[str, str | None]
    signal_type: str
    side: str
    generated_at: datetime
    bar_timestamp: datetime
    quantity: str | None
    target_weight: str | None
    reference_price: str | None
    confidence: str | None
    reason: str | None
    schema_version: int


class StrategySignalPageResponse(BaseModel):
    items: list[StrategySignalResponse]
    page: int
    page_size: int
    total: int


class StrategyExperimentCreateBody(BaseModel):
    strategy_key: str = Field(min_length=2, max_length=64)
    parameter_grid: dict[str, list[str | int | bool]]
    instrument_ids: list[UUID] = Field(min_length=1, max_length=100)
    timeframe: str
    start_at: datetime
    end_at: datetime
    price_adjustment_mode: PriceAdjustmentMode = PriceAdjustmentMode.RAW
    idempotency_key: str = Field(min_length=1, max_length=128)


class StrategyExperimentResponse(BaseModel):
    experiment_id: UUID
    idempotency_key: str
    strategy_key: str
    strategy_version: str
    environment: str
    timeframe: str
    instrument_ids: list[UUID]
    start_at: datetime
    end_at: datetime
    parameter_grid: dict[str, list[str | int | bool]]
    combination_count: int
    runs_completed: int
    runs_failed: int
    total_signals: int
    status: str
    started_at: datetime | None
    completed_at: datetime | None
    failed_at: datetime | None
    error: dict[str, str] | None
    correlation_id: UUID
    created_at: datetime
    updated_at: datetime
    replayed: bool = False
    capabilities: dict[str, bool]


class StrategyExperimentPageResponse(BaseModel):
    items: list[StrategyExperimentResponse]
    page: int
    page_size: int
    total: int


class StrategyExperimentRunResponse(BaseModel):
    combination_index: int
    normalized_parameters: dict[str, str | int | bool]
    strategy_run_id: UUID
    run_status: str
    bars_processed: int
    signals_generated: int
    warning: str | None


class StrategyExperimentComparisonResponse(BaseModel):
    combination_index: int
    normalized_parameters: dict[str, str | int | bool]
    strategy_run_id: UUID
    run_status: str
    bars_processed: int
    total_signals: int
    buy_signals: int
    sell_signals: int
    first_signal_at: datetime | None
    last_signal_at: datetime | None
    signaled_instrument_count: int
    warning: str | None


class StrategySignalOverlapResponse(BaseModel):
    left_combination_index: int
    right_combination_index: int
    intersection_count: int
    union_count: int
    similarity: str
